pkg/util/taints/taints_test.go

```diff
@@ -106,6 +154,118 @@ func TestTaintExists(t *testing.T) {
 	}
 }
 
+func TestTaintKeyExists(t *testing.T) {
+	testingTaints := []v1.Taint{
+		{
+			Key:    "foo_1",
+			Value:  "bar_1",
+			Effect: v1.TaintEffectNoExecute,
+		},
+		{
+			Key:    "foo_2",
+			Value:  "bar_2",
+			Effect: v1.TaintEffectNoSchedule,
+		},
+	}
+
+	cases := []struct {
+		name            string
+		taintKeyToMatch string
+		expectedResult  bool
+	}{
+		{
+			name:            "taint key exists",
+			taintKeyToMatch: "foo_1",
+			expectedResult:  true,
+		},
+		{
+			name:            "taint key does not exist",
+			taintKeyToMatch: "foo_3",
+			expectedResult:  false,
+		},
+	}
+
+	for _, c := range cases {
+		result := TaintKeyExists(testingTaints, c.taintKeyToMatch)
+
+		if result != c.expectedResult {
+			t.Errorf("[%s] unexpected results: %v", c.name, result)
+			continue
+		}
+	}
+}
+
+func TestTaintSetFilter(t *testing.T) {
+	testTaint1 := v1.Taint{
+		Key:    "foo_1",
+		Value:  "bar_1",
+		Effect: v1.TaintEffectNoExecute,
+	}
+	testTaint2 := v1.Taint{
+		Key:    "foo_2",
+		Value:  "bar_2",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	testTaint3 := v1.Taint{
+		Key:    "foo_3",
+		Value:  "bar_3",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+	testingTaints := []v1.Taint{testTaint1, testTaint2, testTaint3}
+
+	testcases := []struct {
+		name           string
+		taints         []v1.Taint
+		fn             func(t *v1.Taint) bool
+		expectedTaints []v1.Taint
+	}{
+		{
+			name:   "filter out 0 taint",
+			taints: testingTaints,
+			fn: func(t *v1.Taint) bool {
+				if t.Key == v1.TaintNodeUnschedulable {
+					return true
+				}
+				return false
+			},
+			expectedTaints: []v1.Taint{},
+		},
+		{
+			name:   "filter out 1 taint",
+			taints: testingTaints,
+			fn: func(t *v1.Taint) bool {
+				if t.Effect == v1.TaintEffectNoExecute {
+					return true
+				}
+				return false
+			},
+			expectedTaints: []v1.Taint{testTaint1},
+		},
+		{
+			name:   "filter out 2 taint",
+			taints: testingTaints,
+			fn: func(t *v1.Taint) bool {
+				if t.Effect == v1.TaintEffectNoSchedule {
+					return true
+				}
+				return false
+			},
+			expectedTaints: []v1.Taint{testTaint2, testTaint3},
+		},
+		{
+			name:           "do not filter",
+			taints:         testingTaints,
+			fn:             func(t *v1.Taint) bool { return true },
+			expectedTaints: []v1.Taint{testTaint1, testTaint2, testTaint3},
+		},
+	}
+	for _, tc := range testcases {
+		taintsAfterFilter := TaintSetFilter(tc.taints, tc.fn)
+		assert.Equal(t, tc.expectedTaints, taintsAfterFilter)
```

Comment:
Is there a reason to use `asset.Equal` here instead of `reflect.DeepEqual` as done before? Doing it in the latter way will allow us to log test names as well as done in the other tests.

Reply:
`assert.equal` output on failure:
```
=== RUN   TestTaintSetFilter
    taints_test.go:240: 
        	Error Trace:	taints_test.go:240
        	Error:      	Not equal: 
        	            	expected: []v1.Taint{}
        	            	actual  : []v1.Taint{v1.Taint{Key:"foo_1", Value:"bar_1", Effect:"NoExecute", TimeAdded:<nil>}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,8 @@
        	            	-([]v1.Taint) {
        	            	+([]v1.Taint) (len=1) {
        	            	+ (v1.Taint) {
        	            	+  Key: (string) (len=5) "foo_1",
        	            	+  Value: (string) (len=5) "bar_1",
        	            	+  Effect: (v1.TaintEffect) (len=9) "NoExecute",
        	            	+  TimeAdded: (*v1.Time)(<nil>)
        	            	+ }
        	            	 }
        	Test:       	TestTaintSetFilter
--- FAIL: TestTaintSetFilter (0.00s)


Expected :[]v1.Taint{}
Actual   :[]v1.Taint{v1.Taint{Key:"foo_1", Value:"bar_1", Effect:"NoExecute", TimeAdded:<nil>}}
<Click to see difference>


FAIL
```

reflect.DeepEqual output on failure:
```
=== RUN   TestTaintSetFilter
    taints_test.go:240: [Filter out a subset ] expect [], but got: []v1.Taint{v1.Taint{Key:"foo_1", Value:"bar_1", Effect:"NoExecute", TimeAdded:<nil>}}
--- FAIL: TestTaintSetFilter (0.00s)
```

Therefore, using the original method `reflect.DeepEqual` is more concise and more consistent with the previous error message. I will update it. Thanks!